src/consensus/encode.rs

```diff
@@ -547,6 +547,38 @@ impl Encodable for [u16; 8] {
     }
 }
 
+/// Read wrapper capped to `remaining` bytes (default: [MAX_VEC_SIZE])
+pub struct CappedRead<'r> {
+    reader: &'r mut dyn io::Read,
+    remaining: usize,
+}
+
+impl<'r> CappedRead<'r> {
+    /// New [CappedRead] from a Read type capped to [MAX_VEC_SIZE] bytes
+    pub fn new<R: io::Read>(reader: &'r mut R) -> Self {
+        Self::with_cap(reader, MAX_VEC_SIZE)
+    }
+
+    /// New [CappedRead] from a Read type with custom max bytes
+    pub fn with_cap<R: io::Read>(reader: &'r mut R, cap: usize) -> Self {
+        CappedRead {
+            reader: reader as &'r mut dyn io::Read,
+            remaining: cap,
+        }
+    }
+}
+
+impl<'r> io::Read for CappedRead<'r> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let bytes = self.reader.read(buf)?;
+        self.remaining = self.remaining
+            .checked_sub(bytes)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "cap read exceeded"))?;
```

Comment:
`cap read` is not very readable. I imagine eventually some unfortunate non-technical users will be scratching their head looking at it. IMO "while reading, maximum allowed message size exceeded" would be better. Possibly with the limit in the error itself put into it.

Reply:
true, but not applicable anymore because we now use `take()`